CA-rvinson/LaserRunnerRemoval
# Unity simulator

Unity-based simulator for the Laser Runner Cutter. ROS2 communication from Unity is done via [ros2-for-unity](https://github.com/RobotecAI/ros2-for-unity). See README.md.

## Alternatives Considered

I looked into both Gazebo and Unity and ended up choosing Unity for the following reasons:
1. Familiarity. Gazebo's documentation is quite lacking and also has a steep learning curve
2. Better graphics capabilities (highly customizeable graphics pipeline, shader support, asset import, etc.) on Unity which will allow photorealistic rendering in the future.
3. Good support for ROS. In particular, there are two commonly used solutions: the official ROS2 connector (https://github.com/Unity-Technologies/ROS-TCP-Connector) and https://github.com/RobotecAI/ros2-for-unity. However, the official ROS2 connector written by Unity does not seem to be actively worked on and also has non-negligible latency (communication bridge over TCP). ros2-for-unity on the other hand is as close to native communication as you can get with Unity, and contains all the features we need.

There are downsides of using Unity that don't really affect us:
1. Not open source. That's fine, as we're not looking to modify the game engine.
2. Physics engine (Nvidia PhysX) is non-deterministic and not quantitatively accurate (most notably, the friction model). We don't depend on very high accuracy in physics simulation anyway, so it's not a deal breaker. If we do need better physics simulation, there are more accurate, deterministic engines (Unity Physics, Havok) if we switch to using DOTS.

## Test Plan

1. Open Unity project
5. Open and run the LaserRunnerCutterScene scene
6. Run rqt, open the ImageView plugin, and listen to the /color_frame topic or /depth_frame topic
7. Verify that the camera image appears in rqt
